### PR TITLE
2500 floodsub -- WIP

### PR DESCRIPTION
**Pull request by nloadholtes**
# Description

Just for me to be able to figure out what I've changed

## Change checklist

- [ ] I have performed a self-review of my own code
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] A changelog entry has been made in the appropriate crates

### CODE REVIEW

File: protocols/floodsub/src/protocol.rs [hunk 1]
```diff
@@ -28,14 +28,42 @@ use libp2p_core::{upgrade, InboundUpgrade, OutboundUpgrade, PeerId, UpgradeInfo}
...
 
 impl FloodsubProtocol {
     /// Builds a new `FloodsubProtocol`.
     pub fn new() -> FloodsubProtocol {
-        FloodsubProtocol {}
+        let codec = prost_codec::Codec::new(MAX_MESSAGE_LEN_BYTES);
+        FloodsubProtocol { codec }
+    }
+}
+
```
Comment (nloadholtes) [inline protocols/floodsub/src/protocol.rs]: These 3 impl's were added because the `#[derive(Debug, Clone, Default)]`. When that was there it complained the Codec didn't implement those. When I added these impls, then the need for the `Derive` went away.

I'm not sure this is the appropriate approach: It seems like Codec should handle this (e.g. it wasn't a problem in `libp2p-gossipsub`)

File: protocols/floodsub/src/protocol.rs [hunk 2]
```diff
@@ -58,8 +86,10 @@ where
 
     fn upgrade_inbound(self, mut socket: TSocket, _: Self::Info) -> Self::Future {
         Box::pin(async move {
-            let packet = upgrade::read_length_prefixed(&mut socket, 2048).await?;
-            let rpc = rpc_proto::Rpc::decode(&packet[..]).map_err(DecodeError)?;
+            let packet = upgrade::read_length_prefixed(&mut socket, MAX_MESSAGE_LEN_BYTES).await?;
+            // Replace with prost_codec::Codec
+            let rpc = self.codec.decode();
```
Comment (nloadholtes) [inline protocols/floodsub/src/protocol.rs]: @thomaseizinger (When you've got a minute, I've got a rust question that I can't seem to find an answer for)
 
The compiler complains about this line and it doesn't make sense to me. The clippy error is:

```
    |
91  |             let rpc = self.codec.decode();
    |                                  ^^^^^^ this is an associated function, not a method
    |
```
It offers suggestions to "fix" this, but those suggestions just lead to more errors/complaints.

My confusion: `self` is identified as a `FloodsubProtocol` object. `codec` is a field there, and the `new()` is defined and sets `codec`... so it seems to me like there's an initialized field there.

What have I missed?

Comment (thomaseizinger) [reply protocols/floodsub/src/protocol.rs]: What you probably want to do here is use a `Framed` to wrap the incoming socket and immediately read a message from it using its `Stream::next` implementation.

`Framed` is an abstraction on top of a codec to read and write from a socket in entire messages instead of raw bytes!

File: protocols/floodsub/src/protocol.rs [hunk 3]
```diff
@@ -28,14 +28,42 @@ use libp2p_core::{upgrade, InboundUpgrade, OutboundUpgrade, PeerId, UpgradeInfo}
...
+            codec: prost_codec::Codec::new(MAX_MESSAGE_LEN_BYTES),
+        }
+    }
+}
+
+// The current error: Maybe if this was in prost_codec everyone would be happier
+impl Clone for FloodsubProtocol {
+    fn clone(&self) -> Self {
+        Self {
+            // TODO: No. This is not the way.
```
Comment (thomaseizinger) [inline protocols/floodsub/src/protocol.rs]: Should be fine actually :)

File: protocols/floodsub/src/protocol.rs (hunk 1 above)
Comment (thomaseizinger) [reply protocols/floodsub/src/protocol.rs]: Yeah this is fine!

Agree that `Codec` should probably implement `Debug`.

File: protocols/floodsub/src/protocol.rs [hunk 4]
```diff
@@ -58,8 +88,20 @@ where
...
+            //
+            // Framed is an abstraction on top of a codec to
+            // read and write from a socket in entire messages instead of
+            // raw bytes!
+
+            Framed::new(socket, FloodsubProtocol::new());
+            let packet = upgrade::read_length_prefixed(&mut socket, MAX_MESSAGE_LEN_BYTES).await?;
+            // Replace with prost_codec::Codec
+            let rpc = self.codec.decode();
+            // prost_codec::Codec::<rpc_proto::Rpc>::decode(&packet[..]).map_err(DecodeError)?;
```
Comment (thomaseizinger) [inline protocols/floodsub/src/protocol.rs]:
```suggestion
            let mut framed = Framed::new(socket, FloodsubProtocol::new());
            let rpc = framed.next().await; // TODO Needs error handling
```

This should do the trick!

File: protocols/floodsub/src/protocol.rs [hunk 5]
```diff
@@ -105,6 +147,7 @@ pub enum FloodsubDecodeError {
     InvalidPeerId,
 }
 
+// TODO: Delete this and change references to prost_codec::Error?
```
Comment (thomaseizinger) [inline protocols/floodsub/src/protocol.rs]: Just change the inner error, this wrapper stops the dependency from leaking out. Major semver bumps would require a major bump of this lib if we didn't wrap the error here.

File: protocols/floodsub/src/protocol.rs [hunk 6]
```diff
@@ -20,22 +20,52 @@
...
+
+impl Default for FloodsubProtocol {
+    fn default() -> Self {
+        Self {
+            codec: prost_codec::Codec::new(MAX_MESSAGE_LEN_BYTES),
+        }
+    }
+}
+
+// The current error: Maybe if this was in prost_codec everyone would be happier
```
Comment (thomaseizinger) [inline protocols/floodsub/src/protocol.rs]: That would make sense yeah!

File: protocols/floodsub/src/protocol.rs (hunk 2 above)
Comment (nloadholtes) [reply protocols/floodsub/src/protocol.rs]: This is awesome, thank you! I'm still working my way through it, work has kept me busier-than-usual this week.

File: protocols/floodsub/src/protocol.rs [hunk 7]
```diff
@@ -58,8 +111,18 @@ where
...
+            // What you probably want to do here is use a Framed to wrap
+            // the incoming socket and immediately read a message from it
+            // using its Stream::next implementation.
+            //
+            // Framed is an abstraction on top of a codec to
+            // read and write from a socket in entire messages instead of
+            // raw bytes!
+
+            let mut framed =
+                Framed::<TSocket, FloodsubProtocol>::new(socket, FloodsubProtocol::new());
```
Comment (thomaseizinger) [inline protocols/floodsub/src/protocol.rs]:
```suggestion
                Framed::<TSocket, prost_codec<rpc_proto::Rpc>>::new(socket, prost_codec::Codec::new(MAX_MESSAGE_LEN_BYTES));
```

We can probably do this actually and delete:

- The `Encoder` / `Decoder` impls
- Manual `Clone`, `Debug` and `Default` impls

**Comment by nloadholtes** [comment]
@thomaseizinger I was doing some work on this today and after being stuck for a while I stepped back and had a realization I wanted to run by you:

1) All of the work in this PR is being done under the umbrella of `FloodsubProtocol`.
2) In Gossipsub there seems to be an "extra" struct in the mix: `GossipsubCodec` in addition to `ProtocolConfig`.
3) `FloodsubProtocol` and `ProtocolConfig` both implement `InboundUpgrade` and `OutboundUpgrade`

This has me thinking: should Floodsub look more like Gossipsub?

I'm tempted to just implement it and see, but I wanted to check with in case there's some arch reason the two are slightly different (and if Floodsub's approach is preferred for some reason).

(I made another WIP PR where I cleaned up things and put comments in at the lines that are giving me problems, but I thought I'd ask first before digging the hole deeper. :wink: )

**Comment by thomaseizinger** [comment]
> @thomaseizinger I was doing some work on this today and after being stuck for a while I stepped back and had a realization I wanted to run by you:
> 
>     1. All of the work in this PR is being done under the umbrella of `FloodsubProtocol`.
> 
>     2. In Gossipsub there seems to be an "extra" struct in the mix: `GossipsubCodec` in addition to `ProtocolConfig`.
> 
>     3. `FloodsubProtocol` and `ProtocolConfig` both implement `InboundUpgrade` and `OutboundUpgrade`

Good observation!

The abstraction hierarchy is like this:

- `NetworkBehaviour` constructs 1 `ConnectionHandler` per connection
- `ConnectionHandler` constructs 1 protocol (`ProtocolInfo` + `Inbound/OutboundUpgrade`) per substream

In gossipsub, we have additional logic that controls, what exact protocol we need to negotiate. Hence, the "protocol" struct there is stateful.

In floodsub, such logic does not exist and hence the protocol can be constructed adhoc and does not have any fields. You can therefore also construct the `Framed` adhoc in the `InboundUpgrade` and `OutboundUpgrade` implementations and then decode the messages. That is what I suggested in https://github.com/nloadholtes/rust-libp2p/pull/2#discussion_r1025933399.

File: protocols/floodsub/src/protocol.rs [hunk 8]
```diff
@@ -56,10 +62,16 @@ where
...
-            let packet = upgrade::read_length_prefixed(&mut socket, 2048).await?;
-            let rpc = rpc_proto::Rpc::decode(&packet[..]).map_err(DecodeError)?;
+            let mut framed = Framed::<TSocket, prost_codec::Codec<rpc_proto::Rpc>>::new(
+                socket,
+                prost_codec::Codec::new(MAX_MESSAGE_LEN_BYTES),
+            );
+            let rpc = match framed.next().await {
+                Some(v) => v.unwrap_or_default(),
+                _ => panic!(), // Better error here?
+            };
```
Comment (nloadholtes) [inline protocols/floodsub/src/protocol.rs]: @thomaseizinger I think I'm really close to the answer on this one, but everything I try looks incorrect (or the compiler complains loudly).

I'm drawing a blank on getting the error in the match. I thought it would be `Err(err) => DecodeError;` but the complaints are loud on that one. It claims it wants to match on `rpc_proto::Error` but including that leads to more errors. :thinking:

File: protocols/floodsub/src/protocol.rs [hunk 9]
```diff
@@ -20,16 +20,22 @@
...
+
 use libp2p_core::{upgrade, InboundUpgrade, OutboundUpgrade, PeerId, UpgradeInfo};
 use prost::Message;
 use std::{io, iter, pin::Pin};
 
+const MAX_MESSAGE_LEN_BYTES: usize = 2048;
+
 /// Implementation of `ConnectionUpgrade` for the floodsub protocol.
-#[derive(Debug, Clone, Default)]
+#[derive(Default, Clone)]
```
Comment (thomaseizinger) [inline protocols/floodsub/src/protocol.rs]: We can revert this back to the original code now!

File: protocols/floodsub/src/protocol.rs [hunk 10]
```diff
@@ -56,10 +62,16 @@ where
...
+    fn upgrade_inbound(self, socket: TSocket, _: Self::Info) -> Self::Future {
         Box::pin(async move {
-            let packet = upgrade::read_length_prefixed(&mut socket, 2048).await?;
-            let rpc = rpc_proto::Rpc::decode(&packet[..]).map_err(DecodeError)?;
+            let mut framed = Framed::<TSocket, prost_codec::Codec<rpc_proto::Rpc>>::new(
+                socket,
+                prost_codec::Codec::new(MAX_MESSAGE_LEN_BYTES),
+            );
+            let rpc = match framed.next().await {
+                Some(v) => v.unwrap_or_default(),
```
Comment (thomaseizinger) [inline protocols/floodsub/src/protocol.rs]: We can't do unwrap-or-default here, but we should map the error up. Multiple possibilities:

- Use `try_next()?` instead of `next`
- Match on the result: `Some(Ok(v))` and have another match branch `Some(Err(e))`.

`try_next` is probably cleaner.

File: protocols/floodsub/src/protocol.rs (hunk 8 above)
Comment (thomaseizinger) [reply protocols/floodsub/src/protocol.rs]: The key is to use `return`.

Almost everything is an expression Rust so if you don't use control flow here, both branches of the `match` need to evaluate to the same type because it will be stored in the `rpc` variable.

Something like this:

```rust
None => return Err(FloodsubDecodeError)
```

Now, there is also a shortcut for that. `Option` has a combinator `ok_or_else` that allows you to map the `None` path to an `Err` which allows you to use `?`:

```rust
let result = stream.next().await.unwrap_or_else(|| FloodsubDecodeError)?;
```

(Examples subject to minor tweaking to properly construct the error but you get the idea.)

File: protocols/floodsub/src/protocol.rs [hunk 11]
```diff
@@ -56,10 +63,28 @@ where
...
+            //     None => return DecodeError,
+            //     // _ => panic!(), // Better error here?
+            // };
+
+            // Better approach:
+            // let rpc = framed.next().await.unwrap_or_else(|| DecodeError);
+            // let rpc_a = framed.next().await.ok_or_else(|| DecodeError); //u nwrap_or_else(|e| (*e).clone());
+            // let rpc = rpc_a.unwrap();
+
+            // TODO: Understand what the ? is doing, and how to unwrap this without using a return (return in match causes errors in the map_err below.)
```
Comment (thomaseizinger) [inline protocols/floodsub/src/protocol.rs]: `unwrap_or_else` maps `Option` to `Result` (Hint: Read its source code!) so `?` just unpacks the `Result` and returns the `Err` if present and evaluates to `Ok` otherwise, same functionality as everywhere really :)

File: protocols/floodsub/src/protocol.rs [hunk 12]
```diff
@@ -56,10 +63,28 @@ where
...
+            let rpc = framed
+                .next()
+                .await
+                .unwrap_or_else(|| FloodsubDecodeError::ReadError)?;
```
Comment (thomaseizinger) [inline protocols/floodsub/src/protocol.rs]: You will have to construct an `io::Error` here to put into `ReadError`. "UnexpectedEOF" is likely the most appropriate.

File: protocols/floodsub/src/protocol.rs [hunk 13]
```diff
@@ -20,16 +20,23 @@
...
+
 use libp2p_core::{upgrade, InboundUpgrade, OutboundUpgrade, PeerId, UpgradeInfo};
 use prost::Message;
 use std::{io, iter, pin::Pin};
 
+const MAX_MESSAGE_LEN_BYTES: usize = 2048;
+
 /// Implementation of `ConnectionUpgrade` for the floodsub protocol.
-#[derive(Debug, Clone, Default)]
+#[derive(Default, Clone)]
```
Comment (thomaseizinger) [inline protocols/floodsub/src/protocol.rs]: Let's undo this change.

File: protocols/floodsub/src/protocol.rs [hunk 14]
```diff
@@ -20,16 +20,23 @@
...
 use crate::topic::Topic;
+use asynchronous_codec::Framed;
+
+use futures::StreamExt;
+use futures::TryStreamExt;
 use futures::{
     io::{AsyncRead, AsyncWrite},
     AsyncWriteExt, Future,
 };
+
```
Comment (thomaseizinger) [inline protocols/floodsub/src/protocol.rs]: We typically don't put empty lines between use statements.

**Review by thomaseizinger — COMMENTED** [review]
Almost there :)

File: protocols/floodsub/src/protocol.rs (hunk 11 above)
Comment (thomaseizinger) [reply protocols/floodsub/src/protocol.rs]: You may need another `?` to get rid of the 2nd `Result`!

File: protocols/floodsub/src/protocol.rs (hunk 12 above)
Comment (thomaseizinger) [inline protocols/floodsub/src/protocol.rs]:
```suggestion
                .unwrap_or_else(|| FloodsubDecodeError::ReadError)?.map_err(DecodeError)?;
```

This may be needed.

File: protocols/floodsub/src/protocol.rs (hunk 11 above)
Comment (nloadholtes) [reply protocols/floodsub/src/protocol.rs]: Ahh, thank you! I was twisted up badly on this one, but I think I've got it now.  Also, reading the docs I think I was using `unwrap_or_else` when I should have been using `ok_or_else`. (The docs make it sound like `unwrap_or_else` should be used if I have a value I can return, and `ok_or_else` when I need to returns/handle an error)

**Comment by nloadholtes** [comment]
@thomaseizinger My apologies for that last commit, I was trying to push a branch between my home and work machine so I could resume working on this. Your tip were spot on, and I think I'm the right track! I'm going to start the `upgrade_outbound` function and see if I can get this back on track :)

File: protocols/floodsub/src/protocol.rs (hunk 11 above)
Comment (thomaseizinger) [reply protocols/floodsub/src/protocol.rs]: My bad yes, it is `ok_or_else`. Sorry for sending you down that path!

**Comment by nloadholtes** [comment]
I think this is ready for the upstream review. Closing this one, will open against upstream
